File: Java/Loon-Neo/src/loon/utils/Calculator.java
/**
 * 
 * Copyright 2008 - 2009
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 * 
 * @project loon
 * @author cping
 * @email：javachenpeng@yahoo.com
 * @version 0.1.1
 */
package loon.utils;

import loon.LRelease;

/**
 * 简单的四则运算类,脚本用,也可用于跨域传参(单纯跨域传参推荐使用 @see Counter 类,因为该类可以限制最大和最小值,更适合简单增减用).
 */
public class Calculator implements LRelease{

	public static final int ADD = 1;

	public static final int SUBTRACT = 2;

	public static final int MULTIPLY = 3;

	public static final int DIVIDE = 4;

	public static final int MODULO = 5;

	public static final int EQUAL = 6;

	private float currentTotal;

	public Calculator() {
		this(0);
	}

	public Calculator(Object num) {
		this(convertObjectToFloat(num));
	}

	public Calculator(float num) {
		this.currentTotal = num;
	}

	protected static final float convertObjectToFloat(Object num) {
		if (num == null) {
			return -1f;
		}
		float value;
		if (num instanceof Number) {
			value = ((Number) num).floatValue();
		} else {
			String mes = num.toString();
			if (MathUtils.isNan(mes)) {
				value = Float.parseFloat(mes);
			} else {
				value = -1f;
			}
		}
		return value;
	}

	public Calculator add(Object num) {
		return add(convertObjectToFloat(num));
	}

	public Calculator sub(Object num) {
		return sub(convertObjectToFloat(num));
	}

	public Calculator mul(Object num) {
		return mul(convertObjectToFloat(num));
	}

	public Calculator div(Object num) {
		return div(convertObjectToFloat(num));
	}

	public Calculator mod(Object num) {
		return mod(convertObjectToFloat(num));
	}

	public Calculator equal(Object num) {
		return div(convertObjectToFloat(num));
	}

	public Calculator add(String num) {
		return convertToFloat(num, ADD);
	}

	public Calculator sub(String num) {
		return convertToFloat(num, SUBTRACT);
	}

	public Calculator mul(String num) {
		return convertToFloat(num, MULTIPLY);
	}

	public Calculator div(String num) {
		return convertToFloat(num, DIVIDE);
	}

	public Calculator mod(String num) {
		return convertToFloat(num, MODULO);
	}

	private Calculator convertToFloat(String num, int operator) {
		if (MathUtils.isNan(num)) {
			float dblNumber = Float.parseFloat(num);
			switch (operator) {
			case ADD:
				return add(dblNumber);
			case SUBTRACT:
				return sub(dblNumber);
			case MULTIPLY:
				return mul(dblNumber);
			case DIVIDE:
				return div(dblNumber);
			case MODULO:
				return mod(dblNumber);
			case EQUAL:
				return equal(dblNumber);
			default:
				break;
			}
		}
		return this;
	}

	public Calculator add(float num) {
		currentTotal += num % 1f == 0 ? (int) num : num;
		return this;
	}
	
	public Calculator sub(float num) {
		currentTotal -= num % 1f == 0 ? (int) num : num;
		return this;
	}

	public Calculator mul(float num) {
		currentTotal *= num % 1f == 0 ? (int) num : num;
		return this;
	}

	public Calculator div(float num) {
		currentTotal /= num % 1f == 0 ? (int) num : num;
		return this;
	}

	public Calculator mod(float num) {
		currentTotal %= num % 1f == 0 ? (int) num : num;
		return this;
	}

	public Calculator equal(float num) {
		currentTotal = num;
		return this;
	}

	public Calculator equal(String num) {
		if (MathUtils.isNan(num)) {
			currentTotal = Float.parseFloat(num);
		}
		return this;
	}

	public int getInt() {
		return (int) currentTotal;
	}

	public float getFloat() {
		return currentTotal;
	}

	@Override
	public String toString() {
		return currentTotal % 1f == 0 ? Integer.toString(getInt()) : String.valueOf(currentTotal);
	}

	@Override
	public void close() {
		this.currentTotal = 0f;
	}

}

File: C#/Loon2MonoGame/LoonMonoGame-Lib/loon/utils/Calculator.cs
using java.lang;

namespace loon.utils
{
    public class Calculator
    {

        public const int ADD = 1;

        public const int SUBTRACT = 2;

        public const int MULTIPLY = 3;

        public const int DIVIDE = 4;

        public const int MODULO = 5;

        public const int EQUAL = 6;

        private float currentTotal;

        public Calculator() : this(0)
        {

        }

        public Calculator(object num) : this(ConvertObjectToFloat(num))
        {

        }

        public Calculator(float num)
        {
            this.currentTotal = num;
        }

        protected static float ConvertObjectToFloat(object num)
        {
            if (num == null)
            {
                return -1f;
            }
            float value;
            if (num is Number number)
            {
                value = number.FloatValue();
            }
            else
            {
                string mes = num.ToString();
                if (MathUtils.IsNan(mes))
                {
                    value = Float.ParseFloat(mes);
                }
                else
                {
                    value = -1f;
                }
            }
            return value;
        }

        public Calculator Add(object num)
        {
            return Add(ConvertObjectToFloat(num));
        }

        public Calculator Sub(object num)
        {
            return Sub(ConvertObjectToFloat(num));
        }

        public Calculator Mul(object num)
        {
            return Mul(ConvertObjectToFloat(num));
        }

        public Calculator Div(object num)
        {
            return Div(ConvertObjectToFloat(num));
        }

        public Calculator Mod(object num)
        {
            return Mod(ConvertObjectToFloat(num));
        }

        public Calculator Equal(object num)
        {
            return Div(ConvertObjectToFloat(num));
        }

        public Calculator Add(string num)
        {
            return ConvertToFloat(num, ADD);
        }

        public Calculator Sub(string num)
        {
            return ConvertToFloat(num, SUBTRACT);
        }

        public Calculator Mul(string num)
        {
            return ConvertToFloat(num, MULTIPLY);
        }

        public Calculator Div(string num)
        {
            return ConvertToFloat(num, DIVIDE);
        }

        public Calculator Mod(string num)
        {
            return ConvertToFloat(num, MODULO);
        }

        private Calculator ConvertToFloat(string num, int oper)
        {
            if (MathUtils.IsNan(num))
            {
                float dblNumber = Float.ParseFloat(num);
                switch (oper)
                {
                    case ADD:
                        return Add(dblNumber);
                    case SUBTRACT:
                        return Sub(dblNumber);
                    case MULTIPLY:
                        return Mul(dblNumber);
                    case DIVIDE:
                        return Div(dblNumber);
                    case MODULO:
                        return Mod(dblNumber);
                    case EQUAL:
                        return Equal(dblNumber);
                    default:
                        break;
                }
            }
            return this;
        }

        public Calculator Add(float num)
        {
            currentTotal += num % 1f == 0 ? (int)num : num;
            return this;
        }

        public Calculator Sub(float num)
        {
            currentTotal -= num % 1f == 0 ? (int)num : num;
            return this;
        }

        public Calculator Mul(float num)
        {
            currentTotal *= num % 1f == 0 ? (int)num : num;
            return this;
        }

        public Calculator Div(float num)
        {
            currentTotal /= num % 1f == 0 ? (int)num : num;
            return this;
        }

        public Calculator Mod(float num)
        {
            currentTotal %= num % 1f == 0 ? (int)num : num;
            return this;
        }

        public Calculator Equal(float num)
        {
            currentTotal = num;
            return this;
        }

        public Calculator Equal(string num)
        {
            if (MathUtils.IsNan(num))
            {
                currentTotal = Float.ParseFloat(num);
            }
            return this;
        }

        public int GetInt()
        {
            return (int)currentTotal;
        }

        public float GetFloat()
        {
            return currentTotal;
        }

        public override string ToString()
        {
            return currentTotal % 1f == 0 ? Integer.ToString(GetInt()) : StringExtensions.ValueOf(currentTotal);
        }
    }
}
